### PR TITLE
Add Powderbarrel and ITNT

### PR DESCRIPTION
**Pull request by serenibyss**
Supersedes #1019

Both of these are added as step 1 of making the Implosion Compressor better/more useful. However they both have uses outside of this.

In addition to the player-facing features, this comes with a (somewhat generalized) API for making different types of explosives

## Powderbarrel
- Slightly larger area than vanilla TNT, and does not require machines beyond Steam era to craft, but purposefully does not have a machine recipe
- Drops all blocks in a 5x5x5 area around it
- Explosion does not destroy item entities
- Cannot be activated by Redstone and cannot be pushed by pistons
- Can be activated by mining, or picked up by sneak-mining
- Has a longer fuse than vanilla TNT, to accommodate for it being done manually
- Could be very useful for some packs in the early game

## ITNT
- Much larger area than vanilla TNT, 7x7x7
- Same as above, except it can be activated by Redstone (and pushed by pistons), and has a much shorter fuse than vanilla TNT

### CODE REVIEW

**Review by ALongStringOfNumbers — COMMENTED** [review]
Should dynamite be changed over to use the new system (looking at the entity stuff specifically, but there might be more)

File: src/main/java/gregtech/common/entities/EntityGTExplosive.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                this.getPosition().add(range, range, range))) {
+            IBlockState state = world.getBlockState(pos);
+
+            if (state.getMaterial() == Material.AIR) continue;
+            if (state.getMaterial() == Material.WATER || state.getMaterial() == Material.LAVA) continue;
+
+            float hardness = state.getBlockHardness(world, pos);
+            float resistance = state.getBlock().getExplosionResistance(player);
+
+            if (hardness >= 0.0f && resistance < 100 && world.isBlockModifiable(player, pos)) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/entities/EntityGTExplosive.java:113]: should this check for if the resistance is more than the strength (or some factor of the strength)? I am not really sure how regular tnt works

**Review by ALongStringOfNumbers — COMMENTED** [review]
I think the powderbarrel needs a different place and break sound. Right now it just sounds like an oak log.

Also, it is a bit weird that ITNT is instant break with hands, but the powderbarrel has a break time.

Also, I have noticed that the TNT explosion is still a little bigger than powderbarrels, contrasting the tooltip.

File: src/main/java/gregtech/common/entities/EntityGTExplosive.java (hunk 1 above)
Comment (serenibyss) [reply src/main/java/gregtech/common/entities/EntityGTExplosive.java:113]: This works differently than normal TNT in that it guarantees drops so it doesn't use the same checks. But this value was chosen to break most blocks but not things like liquid blocks (resistance 100) and obsidian (1200) and similar high explosion resistance blocks

**Comment by serenibyss** [comment]
> I think the powderbarrel needs a different place and break sound. Right now it just sounds like an oak log.

I kept the original sound, it is also crafted using wood so I think it fits fine

> Also, I have noticed that the TNT explosion is still a little bigger than powderbarrels, contrasting the tooltip.

It depends on the block you try to explode, TNT varies but powderbarrel is consistent size no matter the damaged blocks

**Comment by serenibyss** [comment]
> Should dynamite be changed over to use the new system (looking at the entity stuff specifically, but there might be more)

Dynamite renders very differently since its a thrown entity instead of a block that is activated, so it would be hard to combine, and not much benefit to doing so

**Review by ALongStringOfNumbers — APPROVED** [review]

**Comment by ALongStringOfNumbers** [comment]
Just noticed the todo in GTBeeDefinition about the Explosive bee product. Should that be added in this PR?

**Review by ALongStringOfNumbers — APPROVED** [review]

**Pull request merged**